security/src/main/scala/pl/touk/nussknacker/ui/security/api/LoggedUser.scala

```diff
@@ -48,27 +71,55 @@ object LoggedUser {
     }
   }
 
+  implicit class ImpersonationChecking(val user: LoggedUser) extends AnyVal {
+
+    def canImpersonate: Boolean = user match {
+      case _: AdminUser        => true
+      case _: ImpersonatedUser => false
+      case commonUser: CommonUser =>
+        commonUser.globalPermissions
+          .map(_.toLowerCase)
+          .contains(Permission.Impersonate.toString.toLowerCase)
+    }
+
+  }
+
 }
 
 final case class CommonUser(
     id: String,
     username: String,
     categoryPermissions: Map[String, Set[Permission]] = Map.empty,
     globalPermissions: List[GlobalPermission] = Nil
-) extends LoggedUser {
-
-  def categories(permission: Permission): Set[String] = categoryPermissions.collect {
-    case (category, permissions) if permissions contains permission => category
-  }.toSet
+) extends RealLoggedUser {
+  override val isAdmin: Boolean = false
 
   override def can(category: String, permission: Permission): Boolean = {
     categoryPermissions.get(category).exists(_ contains permission)
   }
 
-  override val isAdmin: Boolean = false
+  def categories(permission: Permission): Set[String] = categoryPermissions.collect {
+    case (category, permissions) if permissions contains permission => category
+  }.toSet
+
 }
 
-final case class AdminUser(id: String, username: String) extends LoggedUser {
-  override def can(category: String, permission: Permission): Boolean = true
+final case class AdminUser(id: String, username: String) extends RealLoggedUser {
   override val isAdmin: Boolean                                       = true
+  override def can(category: String, permission: Permission): Boolean = true
+}
+
+final case class ImpersonatedUser(impersonatedUser: RealLoggedUser, impersonatingUser: RealLoggedUser)
+    extends LoggedUser {
+  override val id: String                                             = impersonatedUser.id
+  override val username: String                                       = impersonatedUser.username
+  override val isAdmin: Boolean                                       = impersonatedUser.isAdmin
```

Comment:
WDYT about boy-scouting here a little bit? AFAIS we don't really need the `isAdmin` field. Maybe we can remove it?

Reply:
Yes, it doesn't seem to be needed as a field. I removed it.